File: tests/io_operations/csv_processing/csv_processing.go
package main

import (
	"encoding/csv"
	"encoding/json"
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"
)

type Config struct {
	Parameters Parameters `json:"parameters"`
}

type Parameters struct {
	RowCounts    []int    `json:"row_counts"`
	ColumnCounts []int    `json:"column_counts"`
	Operations   []string `json:"operations"`
	DataTypes    []string `json:"data_types"`
	Iterations   int      `json:"iterations"`
}

type OperationResult struct {
	Success        bool    `json:"success"`
	TimeMs         float64 `json:"time_ms,omitempty"`
	Error          string  `json:"error,omitempty"`
	OutputSize     int     `json:"output_size,omitempty"`
	RowsRead       int     `json:"rows_read,omitempty"`
	OriginalRows   int     `json:"original_rows,omitempty"`
	FilteredRows   int     `json:"filtered_rows,omitempty"`
	AggregatedCols int     `json:"aggregated_columns,omitempty"`
}

type IterationResult struct {
	Iteration  int                        `json:"iteration"`
	DataSize   int                        `json:"data_size"`
	Operations map[string]OperationResult `json:"operations"`
}

type TestCase struct {
	RowCount         int               `json:"row_count"`
	ColumnCount      int               `json:"column_count"`
	DataType         string            `json:"data_type"`
	Operations       []string          `json:"operations"`
	Iterations       []IterationResult `json:"iterations"`
	AvgReadTime      float64           `json:"avg_read_time"`
	AvgWriteTime     float64           `json:"avg_write_time"`
	AvgFilterTime    float64           `json:"avg_filter_time"`
	AvgAggregateTime float64           `json:"avg_aggregate_time"`
}

type Summary struct {
	TotalTests       int     `json:"total_tests"`
	SuccessfulTests  int     `json:"successful_tests"`
	FailedTests      int     `json:"failed_tests"`
	AvgReadTime      float64 `json:"avg_read_time"`
	AvgWriteTime     float64 `json:"avg_write_time"`
	AvgFilterTime    float64 `json:"avg_filter_time"`
	AvgAggregateTime float64 `json:"avg_aggregate_time"`
}

type Results struct {
	StartTime          float64    `json:"start_time"`
	TestCases          []TestCase `json:"test_cases"`
	Summary            Summary    `json:"summary"`
	EndTime            float64    `json:"end_time"`
	TotalExecutionTime float64    `json:"total_execution_time"`
}

func generateCSVData(rows, cols int, dataType string) [][]string {
	rand.Seed(time.Now().UnixNano())
	data := make([][]string, 0, rows+1)

	// Generate headers
	headers := make([]string, cols)
	for i := 0; i < cols; i++ {
		headers[i] = fmt.Sprintf("col_%d", i+1)
	}
	data = append(data, headers)

	// Generate data rows
	for row := 0; row < rows; row++ {
		rowData := make([]string, cols)
		for col := 0; col < cols; col++ {
			var value string
			switch dataType {
			case "numeric":
				value = fmt.Sprintf("%.2f", rand.Float64()*1000)
			case "text":
				length := rand.Intn(11) + 5 // 5-15 characters
				runes := make([]rune, length)
				for i := range runes {
					runes[i] = rune('a' + rand.Intn(26))
				}
				value = string(runes)
			default: // mixed
				switch col % 3 {
				case 0:
					value = strconv.Itoa(rand.Intn(10000) + 1)
				case 1:
					runes := make([]rune, 10)
					for i := range runes {
						runes[i] = rune('a' + rand.Intn(26))
					}
					value = string(runes)
				default:
					value = fmt.Sprintf("%.2f", rand.Float64()*1000)
				}
			}
			rowData[col] = value
		}
		data = append(data, rowData)
	}

	return data
}

func writeCSVToString(data [][]string) string {
	var result strings.Builder
	for _, row := range data {
		result.WriteString(strings.Join(row, ","))
		result.WriteByte('\n')
	}
	return result.String()
}

func readCSVFromString(csvString string) [][]string {
	reader := csv.NewReader(strings.NewReader(csvString))
	records, err := reader.ReadAll()
	if err != nil {
		return nil
	}
	return records
}

func filterCSVData(data [][]string, filterColumn int) [][]string {
	if len(data) < 2 {
		return data
	}

	filtered := [][]string{data[0]} // Keep headers

	for _, row := range data[1:] {
		if len(row) > filterColumn {
			if value, err := strconv.ParseFloat(row[filterColumn], 64); err == nil {
				if value > 500 {
					filtered = append(filtered, row)
				}
			} else if len(row[filterColumn]) > 5 {
				filtered = append(filtered, row)
			}
		}
	}

	return filtered
}

func aggregateCSVData(data [][]string) map[string]map[string]float64 {
	if len(data) < 2 {
		return make(map[string]map[string]float64)
	}

	headers := data[0]
	var numericColumns []int

	// Find numeric columns (check first 5 rows)
	for colIdx := 0; colIdx < len(headers); colIdx++ {
		isNumeric := true
		checkRows := len(data) - 1
		if checkRows > 5 {
			checkRows = 5
		}

		for rowIdx := 1; rowIdx <= checkRows; rowIdx++ {
			if colIdx < len(data[rowIdx]) {
				if _, err := strconv.ParseFloat(data[rowIdx][colIdx], 64); err != nil {
					isNumeric = false
					break
				}
			}
		}
		if isNumeric {
			numericColumns = append(numericColumns, colIdx)
		}
	}

	aggregations := make(map[string]map[string]float64)

	for _, colIdx := range numericColumns {
		colName := headers[colIdx]
		var values []float64

		for _, row := range data[1:] {
			if colIdx < len(row) {
				if value, err := strconv.ParseFloat(row[colIdx], 64); err == nil {
					values = append(values, value)
				}
			}
		}

		if len(values) > 0 {
			sum := 0.0
			min := values[0]
			max := values[0]

			for _, value := range values {
				sum += value
				if value < min {
					min = value
				}
				if value > max {
					max = value
				}
			}

			stats := map[string]float64{
				"sum":   sum,
				"avg":   sum / float64(len(values)),
				"min":   min,
				"max":   max,
				"count": float64(len(values)),
			}

			aggregations[colName] = stats
		}
	}

	return aggregations
}

func contains(slice []string, item string) bool {
	for _, s := range slice {
		if s == item {
			return true
		}
	}
	return false
}

func runCSVProcessingBenchmark(config Config) Results {
	parameters := config.Parameters

	// Set defaults
	rowCounts := parameters.RowCounts
	if len(rowCounts) == 0 {
		rowCounts = []int{1000}
	}

	columnCounts := parameters.ColumnCounts
	if len(columnCounts) == 0 {
		columnCounts = []int{5}
	}

	operations := parameters.Operations
	if len(operations) == 0 {
		operations = []string{"read", "write"}
	}

	dataTypes := parameters.DataTypes
	if len(dataTypes) == 0 {
		dataTypes = []string{"mixed"}
	}

	iterations := parameters.Iterations
	if iterations == 0 {
		iterations = 3
	}

	startTime := time.Now()
	var testCases []TestCase
	var allReadTimes, allWriteTimes, allFilterTimes, allAggregateTimes []float64
	totalTests := 0
	successfulTests := 0
	failedTests := 0

	for _, rows := range rowCounts {
		for _, cols := range columnCounts {
			for _, dataType := range dataTypes {
				fmt.Fprintf(os.Stderr, "Testing CSV: %d rows x %d cols, type: %s...\n", rows, cols, dataType)

				var readTimes, writeTimes, filterTimes, aggregateTimes []float64
				var iterationsData []IterationResult

				for i := 0; i < iterations; i++ {
					fmt.Fprintf(os.Stderr, "  Iteration %d/%d...\n", i+1, iterations)

					// Generate test data
					csvData := generateCSVData(rows, cols, dataType)

					iterationResult := IterationResult{
						Iteration:  i + 1,
						DataSize:   len(csvData),
						Operations: make(map[string]OperationResult),
					}

					totalTests++
					success := true

					// Write operation
					if contains(operations, "write") {
						start := time.Now()
						csvString := writeCSVToString(csvData)
						writeTime := float64(time.Since(start).Nanoseconds()) / 1000000.0

						writeTimes = append(writeTimes, writeTime)
						allWriteTimes = append(allWriteTimes, writeTime)

						iterationResult.Operations["write"] = OperationResult{
							Success:    true,
							TimeMs:     writeTime,
							OutputSize: len(csvString),
						}
					}

					// Read operation
					if contains(operations, "read") {
						csvString := writeCSVToString(csvData)

						start := time.Now()
						readData := readCSVFromString(csvString)
						readTime := float64(time.Since(start).Nanoseconds()) / 1000000.0

						readTimes = append(readTimes, readTime)
						allReadTimes = append(allReadTimes, readTime)

						iterationResult.Operations["read"] = OperationResult{
							Success:  true,
							TimeMs:   readTime,
							RowsRead: len(readData),
						}
					}

					// Filter operation
					if contains(operations, "filter") {
						start := time.Now()
						filteredData := filterCSVData(csvData, 0)
						filterTime := float64(time.Since(start).Nanoseconds()) / 1000000.0

						filterTimes = append(filterTimes, filterTime)
						allFilterTimes = append(allFilterTimes, filterTime)

						iterationResult.Operations["filter"] = OperationResult{
							Success:      true,
							TimeMs:       filterTime,
							OriginalRows: len(csvData),
							FilteredRows: len(filteredData),
						}
					}

					// Aggregate operation
					if contains(operations, "aggregate") {
						start := time.Now()
						aggregations := aggregateCSVData(csvData)
						aggregateTime := float64(time.Since(start).Nanoseconds()) / 1000000.0

						aggregateTimes = append(aggregateTimes, aggregateTime)
						allAggregateTimes = append(allAggregateTimes, aggregateTime)

						iterationResult.Operations["aggregate"] = OperationResult{
							Success:        true,
							TimeMs:         aggregateTime,
							AggregatedCols: len(aggregations),
						}
					}

					if success {
						successfulTests++
					} else {
						failedTests++
					}

					iterationsData = append(iterationsData, iterationResult)
				}

				// Calculate averages for this test case
				testCase := TestCase{
					RowCount:    rows,
					ColumnCount: cols,
					DataType:    dataType,
					Operations:  operations,
					Iterations:  iterationsData,
				}

				if len(readTimes) > 0 {
					sum := 0.0
					for _, t := range readTimes {
						sum += t
					}
					testCase.AvgReadTime = sum / float64(len(readTimes))
				}

				if len(writeTimes) > 0 {
					sum := 0.0
					for _, t := range writeTimes {
						sum += t
					}
					testCase.AvgWriteTime = sum / float64(len(writeTimes))
				}

				if len(filterTimes) > 0 {
					sum := 0.0
					for _, t := range filterTimes {
						sum += t
					}
					testCase.AvgFilterTime = sum / float64(len(filterTimes))
				}

				if len(aggregateTimes) > 0 {
					sum := 0.0
					for _, t := range aggregateTimes {
						sum += t
					}
					testCase.AvgAggregateTime = sum / float64(len(aggregateTimes))
				}

				testCases = append(testCases, testCase)
			}
		}
	}

	// Calculate overall summary
	summary := Summary{
		TotalTests:      totalTests,
		SuccessfulTests: successfulTests,
		FailedTests:     failedTests,
	}

	if len(allReadTimes) > 0 {
		sum := 0.0
		for _, t := range allReadTimes {
			sum += t
		}
		summary.AvgReadTime = sum / float64(len(allReadTimes))
	}

	if len(allWriteTimes) > 0 {
		sum := 0.0
		for _, t := range allWriteTimes {
			sum += t
		}
		summary.AvgWriteTime = sum / float64(len(allWriteTimes))
	}

	if len(allFilterTimes) > 0 {
		sum := 0.0
		for _, t := range allFilterTimes {
			sum += t
		}
		summary.AvgFilterTime = sum / float64(len(allFilterTimes))
	}

	if len(allAggregateTimes) > 0 {
		sum := 0.0
		for _, t := range allAggregateTimes {
			sum += t
		}
		summary.AvgAggregateTime = sum / float64(len(allAggregateTimes))
	}

	endTime := time.Now()
	totalExecutionTime := endTime.Sub(startTime).Seconds()

	return Results{
		StartTime:          float64(startTime.Unix()),
		TestCases:          testCases,
		Summary:            summary,
		EndTime:            float64(endTime.Unix()),
		TotalExecutionTime: totalExecutionTime,
	}
}

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintln(os.Stderr, "Usage: csv_processing <config_file>")
		os.Exit(1)
	}

	configFile := os.Args[1]

	configContent, err := os.ReadFile(configFile)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: Config file '%s' not found: %v\n", configFile, err)
		os.Exit(1)
	}

	var config Config
	if err := json.Unmarshal(configContent, &config); err != nil {
		fmt.Fprintf(os.Stderr, "Error: Invalid JSON in config file: %v\n", err)
		os.Exit(1)
	}

	results := runCSVProcessingBenchmark(config)

	output, err := json.MarshalIndent(results, "", "  ")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error marshaling results: %v\n", err)
		os.Exit(1)
	}

	fmt.Println(string(output))
}

File: tests/io_operations/csv_processing/csv_processing.rs
use std::env;
use std::time::Instant;
use serde_json::{json, Value};
use std::fs;
use std::process;
use rand::Rng;



fn generate_csv_data(rows: usize, cols: usize, data_type: &str) -> Vec<Vec<String>> {
    let mut rng = rand::thread_rng();
    let mut data = Vec::new();
    
    // Headers
    let headers: Vec<String> = (0..cols).map(|i| format!("col_{}", i + 1)).collect();
    data.push(headers);
    
    // Data rows
    for _ in 0..rows {
        let mut row = Vec::new();
        for col in 0..cols {
            let value = match data_type {
                "numeric" => format!("{:.2}", rng.gen_range(0.0..1000.0)),
                "text" => {
                    let len = rng.gen_range(5..16);
                    (0..len).map(|_| (b'a' + rng.gen_range(0..26)) as char).collect()
                },
                "mixed" => {
                    match col % 3 {
                        0 => rng.gen_range(1..10001).to_string(),
                        1 => (0..10).map(|_| (b'a' + rng.gen_range(0..26)) as char).collect(),
                        _ => format!("{:.2}", rng.gen_range(0.0..1000.0)),
                    }
                },
                _ => "default".to_string(),
            };
            row.push(value);
        }
        data.push(row);
    }
    
    data
}

fn write_csv_to_string(data: &[Vec<String>]) -> String {
    data.iter()
        .map(|row| row.join(","))
        .collect::<Vec<String>>()
        .join("\n") + "\n"
}

fn read_csv_from_string(csv_string: &str) -> Vec<Vec<String>> {
    csv_string
        .trim()
        .lines()
        .map(|line| line.split(',').map(|s| s.to_string()).collect())
        .collect()
}

fn filter_csv_data(data: &[Vec<String>], filter_column: usize) -> Vec<Vec<String>> {
    if data.is_empty() || data.len() < 2 {
        return data.to_vec();
    }
    
    let mut filtered_data = vec![data[0].clone()]; // Headers
    
    for row in &data[1..] {
        if row.len() > filter_column {
            // Try numeric filter
            if let Ok(value) = row[filter_column].parse::<f64>() {
                if value > 500.0 {
                    filtered_data.push(row.clone());
                }
            } else if row[filter_column].len() > 5 {
                // String length filter
                filtered_data.push(row.clone());
            }
        }
    }
    
    filtered_data
}

fn aggregate_csv_data(data: &[Vec<String>]) -> Value {
    if data.is_empty() || data.len() < 2 {
        return json!({});
    }
    
    let headers = &data[0];
    let mut numeric_columns = Vec::new();
    
    // Find numeric columns
    for (col_idx, _) in headers.iter().enumerate() {
        let mut is_numeric = true;
        for row in data.iter().take(6).skip(1) { // Check first 5 rows
            if row.len() > col_idx {
                if row[col_idx].parse::<f64>().is_err() {
                    is_numeric = false;
                    break;
                }
            }
        }
        if is_numeric {
            numeric_columns.push(col_idx);
        }
    }
    
    let mut aggregations = serde_json::Map::new();
    
    for &col_idx in &numeric_columns {
        let col_name = &headers[col_idx];
        let mut values = Vec::new();
        
        for row in &data[1..] {
            if row.len() > col_idx {
                if let Ok(value) = row[col_idx].parse::<f64>() {
                    values.push(value);
                }
            }
        }
        
        if !values.is_empty() {
            let sum: f64 = values.iter().sum();
            let avg = sum / values.len() as f64;
            let min = values.iter().fold(f64::INFINITY, |a, &b| a.min(b));
            let max = values.iter().fold(f64::NEG_INFINITY, |a, &b| a.max(b));
            
            aggregations.insert(col_name.clone(), json!({
                "sum": sum,
                "avg": avg,
                "min": min,
                "max": max,
                "count": values.len()
            }));
        }
    }
    
    json!(aggregations)
}

fn run_csv_processing_benchmark() -> Value {
    // Default configuration
    let default_config = json!({
        "parameters": {
            "row_counts": [1000],
            "column_counts": [5], 
            "operations": ["read", "write", "filter", "aggregate"],
            "data_types": ["mixed"],
            "iterations": 3
        }
    });
    
    run_csv_processing_benchmark_with_config(&default_config)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    
    let mut config = json!({
        "parameters": {
            "row_counts": [1000],
            "column_counts": [5], 
            "operations": ["read", "write", "filter", "aggregate"],
            "data_types": ["mixed"],
            "iterations": 3
        }
    });
    
    // Check if arguments were passed (from orchestrator)
    if args.len() >= 2 {
        let config_file = &args[1];
        match fs::read_to_string(config_file) {
            Ok(content) => {
                match serde_json::from_str::<Value>(&content) {
                    Ok(parsed) => {
                        config = parsed;
                    },
                    Err(e) => {
                        eprintln!("Error parsing JSON config file: {}", e);
                        process::exit(1);
                    }
                }
            },
            Err(e) => {
                eprintln!("Error: Config file '{}' not found: {}", config_file, e);
                process::exit(1);
            }
        }
    }
    
    let results = run_csv_processing_benchmark_with_config(&config);
    println!("{}", serde_json::to_string_pretty(&results).unwrap());
}


fn run_csv_processing_benchmark_with_config(config: &Value) -> Value {
    // Extract configuration parameters - fix the temporary value issue
    let default_params = json!({});
    let parameters = config.get("parameters").unwrap_or(&default_params);
    
    let row_counts = parameters.get("row_counts")
        .unwrap_or(&json!([1000]))
        .as_array()
        .unwrap_or(&vec![json!(1000)])
        .iter()
        .map(|v| v.as_i64().unwrap_or(1000) as usize)
        .collect::<Vec<usize>>();
    
    let column_counts = parameters.get("column_counts")
        .unwrap_or(&json!([5]))
        .as_array()
        .unwrap_or(&vec![json!(5)])
        .iter()
        .map(|v| v.as_i64().unwrap_or(5) as usize)
        .collect::<Vec<usize>>();
    
    let operations = parameters.get("operations")
        .unwrap_or(&json!(["read", "write", "filter", "aggregate"]))
        .as_array()
        .unwrap_or(&vec![json!("read"), json!("write"), json!("filter"), json!("aggregate")])
        .iter()
        .map(|v| v.as_str().unwrap_or("").to_string())
        .collect::<Vec<String>>();
    
    let data_types = parameters.get("data_types")
        .unwrap_or(&json!(["mixed"]))
        .as_array()
        .unwrap_or(&vec![json!("mixed")])
        .iter()
        .map(|v| v.as_str().unwrap_or("mixed").to_string())
        .collect::<Vec<String>>();
    
    let iterations = parameters.get("iterations")
        .unwrap_or(&json!(1))  // Change to 1 iteration to match the command
        .as_i64()
        .unwrap_or(1) as usize;
    
    
    let start_time = std::time::Instant::now();

    let mut test_cases = Vec::new();
    let mut total_tests = 0;
    let mut successful_tests = 0;
    let mut all_read_times = Vec::new();
    let mut all_write_times = Vec::new();
    let mut all_filter_times = Vec::new();
    let mut all_aggregate_times = Vec::new();

    for &rows in &row_counts {
        for &cols in &column_counts {
            for data_type in &data_types {
                eprintln!("Testing CSV: {} rows x {} cols, type: {}...", rows, cols, data_type);

                let mut iteration_results = Vec::new();
                let mut read_times = Vec::new();
                let mut write_times = Vec::new();
                let mut filter_times = Vec::new();
                let mut aggregate_times = Vec::new();

                for i in 0..iterations {
                    eprintln!("  Iteration {}/{}...", i + 1, iterations);

                    // Generate test data
                    let csv_data = generate_csv_data(rows, cols, data_type);

                    let mut iteration_result = json!({
                        "iteration": i + 1,
                        "data_size": csv_data.len(),
                        "operations": {}
                    });

                    total_tests += 1;
                    let mut success = true;

                    // Write operation
                    if operations.contains(&"write".to_string()) {
                        let start = std::time::Instant::now();
                        let csv_string = write_csv_to_string(&csv_data);
                        let write_time = start.elapsed().as_secs_f64() * 1000.0;

                        write_times.push(write_time);
                        all_write_times.push(write_time);

                        iteration_result["operations"]["write"] = json!({
                            "success": true,
                            "time_ms": write_time,
                            "output_size": csv_string.len()
                        });
                    }

                    // Read operation
                    if operations.contains(&"read".to_string()) {
                        let csv_string = write_csv_to_string(&csv_data);

                        let start = std::time::Instant::now();
                        let read_data = read_csv_from_string(&csv_string);
                        let read_time = start.elapsed().as_secs_f64() * 1000.0;

                        read_times.push(read_time);
                        all_read_times.push(read_time);

                        iteration_result["operations"]["read"] = json!({
                            "success": true,
                            "time_ms": read_time,
                            "rows_read": read_data.len()
                        });
                    }

                    // Filter operation
                    if operations.contains(&"filter".to_string()) {
                        let start = std::time::Instant::now();
                        let filtered_data = filter_csv_data(&csv_data, 0);
                        let filter_time = start.elapsed().as_secs_f64() * 1000.0;

                        filter_times.push(filter_time);
                        all_filter_times.push(filter_time);

                        iteration_result["operations"]["filter"] = json!({
                            "success": true,
                            "time_ms": filter_time,
                            "original_rows": csv_data.len(),
                            "filtered_rows": filtered_data.len()
                        });
                    }

                    // Aggregate operation
                    if operations.contains(&"aggregate".to_string()) {
                        let start = std::time::Instant::now();
                        let aggregations = aggregate_csv_data(&csv_data);
                        let aggregate_time = start.elapsed().as_secs_f64() * 1000.0;

                        aggregate_times.push(aggregate_time);
                        all_aggregate_times.push(aggregate_time);

                        let aggregated_columns = if let Some(obj) = aggregations.as_object() {
                            obj.len()
                        } else {
                            0
                        };

                        iteration_result["operations"]["aggregate"] = json!({
                            "success": true,
                            "time_ms": aggregate_time,
                            "aggregated_columns": aggregated_columns
                        });
                    }

                    if success {
                        successful_tests += 1;
                    }

                    iteration_results.push(iteration_result);
                }

                // Calculate averages
                let avg_read_time = if !read_times.is_empty() {
                    read_times.iter().sum::<f64>() / read_times.len() as f64
                } else { 0.0 };

                let avg_write_time = if !write_times.is_empty() {
                    write_times.iter().sum::<f64>() / write_times.len() as f64
                } else { 0.0 };

                let avg_filter_time = if !filter_times.is_empty() {
                    filter_times.iter().sum::<f64>() / filter_times.len() as f64
                } else { 0.0 };

                let avg_aggregate_time = if !aggregate_times.is_empty() {
                    aggregate_times.iter().sum::<f64>() / aggregate_times.len() as f64
                } else { 0.0 };

                let test_case = json!({
                    "row_count": rows,
                    "column_count": cols,
                    "data_type": data_type,
                    "operations": operations.iter().map(|s| s.as_str()).collect::<Vec<&str>>(),
                    "iterations": iteration_results,
                    "avg_read_time": avg_read_time,
                    "avg_write_time": avg_write_time,
                    "avg_filter_time": avg_filter_time,
                    "avg_aggregate_time": avg_aggregate_time
                });

                test_cases.push(test_case);
            }
        }
    }

    // Calculate overall summary
    let avg_read_time = if !all_read_times.is_empty() {
        all_read_times.iter().sum::<f64>() / all_read_times.len() as f64
    } else { 0.0 };

    let avg_write_time = if !all_write_times.is_empty() {
        all_write_times.iter().sum::<f64>() / all_write_times.len() as f64
    } else { 0.0 };

    let avg_filter_time = if !all_filter_times.is_empty() {
        all_filter_times.iter().sum::<f64>() / all_filter_times.len() as f64
    } else { 0.0 };

    let avg_aggregate_time = if !all_aggregate_times.is_empty() {
        all_aggregate_times.iter().sum::<f64>() / all_aggregate_times.len() as f64
    } else { 0.0 };

    let total_duration = start_time.elapsed().as_secs_f64();

    json!({
        "start_time": 0.0,
        "test_cases": test_cases,
        "summary": {
            "total_tests": total_tests,
            "successful_tests": successful_tests,
            "failed_tests": total_tests - successful_tests,
            "avg_read_time": avg_read_time,
            "avg_write_time": avg_write_time,
            "avg_filter_time": avg_filter_time,
            "avg_aggregate_time": avg_aggregate_time
        },
        "end_time": total_duration,
        "total_execution_time": total_duration
    })
}